code/obj/artifacts/artifact_machines/heater.dm

```diff
@@ -1,47 +1,111 @@
+				gas_type = pick(
+					100;"nitrogen",
+					100;"plasma",
+					100;"carbon dioxide",
+					75;"sleeping agent")
+			if("martian") // organic stuff
+				gas_type = pick(
+					200;"oxygen",
+					100;"nitrogen",
+					50;"carbon dioxide",
+					50;"farts")
+			if("ancient") // industrial type stuff
+				gas_type = pick(
+					200;"oxygen",
+					100;"nitrogen",
+					50;"carbon dioxide",
+					50;"agent b")
+			if("precursor") // the rest
+				gas_type = pick(
+					125;"oxygen",
+					100;"nitrogen",
+					75;"plasma",
+					50;"carbon dioxide",
+					30;"farts",
+					30;"agent b",
+					30;"sleeping agent")
+
+		// temperature
+		gas_temp = rand(0,620)
 		if (artitype.name == "eldritch" && prob(66))
-			if (heat_target > 310)
-				heat_target *= 2
-			if (heat_target < 310 && heat_target != 0)
-				heat_target /= 2
-		heat_amount = rand(20000,60000)
-		if (heat_target > 310)
-			activ_text = "begins to radiate hot air!"
-			deact_text = "stops pumping out hot air."
-		if (heat_target < 310)
-			activ_text = "begins to radiate cold air!"
-			deact_text = "stops pumping out cold air."
+			if (gas_temp > 310)
+				gas_temp *= 2
+			if (gas_temp < 310)
+				gas_temp /= 2
+
+		// amount
+		gas_amount = rand(50,200)
+
+		// text
+		var/tempText = ""
+		if (gas_temp > 310)
+			tempText = "hot "
+		else if (gas_temp < 310)
+			tempText = "cold "
+
+		src.activ_text = "begins to emit [tempText] gas!"
+		src.deact_text = "stops emitting [tempText] gas."
 
 	effect_process(var/obj/O)
 		if (..())
 			return
+		if(src.gas_amount_current < src.gas_amount)
+			src.gas_amount_current = min(src.gas_amount_current + src.gas_amount_growth, src.gas_amount)
 		var/turf/simulated/L = get_turf(O)
 		if(istype(L))
-			var/datum/gas_mixture/env = L.return_air()
-			if(env.temperature < (heat_target+T0C))
-				var/transfer_moles = 0.25 * TOTAL_MOLES(env)
-				var/datum/gas_mixture/removed = env.remove(transfer_moles)
-				if(removed)
-					var/heat_capacity = HEAT_CAPACITY(removed)
-					if(heat_capacity)
-						removed.temperature = (removed.temperature*heat_capacity + heat_amount)/heat_capacity
-				env.merge(removed)
+			var/datum/gas_mixture/gas = unpool(/datum/gas_mixture)
+			gas.zero()
+			switch(src.gas_type)
+				if("oxygen")
+					gas.oxygen = src.gas_amount_current
+				if("nitrogen")
+					gas.nitrogen = src.gas_amount_current
+				if("plasma")
+					gas.toxins = src.gas_amount_current
+				if("carbon dioxide")
+					gas.carbon_dioxide = src.gas_amount_current
+				if("farts")
+					gas.farts = src.gas_amount_current
+				if("agent b")
+					var/datum/gas/oxygen_agent_b/trace = gas.get_or_add_trace_gas_by_type(/datum/gas/oxygen_agent_b)
+					trace.moles = src.gas_amount_current
+				if("sleeping agent")
+					var/datum/gas/sleeping_agent/trace = gas.get_or_add_trace_gas_by_type(/datum/gas/sleeping_agent)
+					trace.moles = src.gas_amount_current
+			gas.temperature = src.gas_temp
+			gas.volume = R_IDEAL_GAS_EQUATION * src.gas_temp / 1000
```

Comment:
Why div 1000

Reply:
Because a bunch of other places also had it, do I look like I know how physics work?^^

Reply:
ok, I asked on imcoder and supposedly the volume does not matter if it's merged right after anyway?
So I'll just remove this, I guess. o.o